### PR TITLE
Local Dependencies

### PR DESCRIPTION
**Pull request by mara-schulke**
This introduces the ability to do the following in `Proto.toml` files:

```
[dependencies]
some-api = { path = "./path/to/some-api" }
```

## Notes

- Publishing any project with local dependencies is *not* possible
- This should be used primarily / only in consuming projects -> e.g. server impls
- This MR bumps the edition to `0.9.0`

### CODE REVIEW

File: src/resolver.rs [hunk 1]
```diff
@@ -1,31 +1,64 @@
...
+        depends_on: Vec<PackageName>,
+    },
+    /// A resolved depnedency that is located on the filesystem
+    Local {
+        /// The materialized package that was created from the buffrs package at the given path
+        package: Package,
+        /// Location of the requested package
+        path: PathBuf,
+        /// Packages that requested this dependency (and what versions they accept)
+        dependants: Vec<Dependant>,
```
Comment (Tpt) [inline src/resolver.rs:42]: What about just a `Vec<String>`? I believe the version range is always going to be `*`

File: src/resolver.rs [hunk 2]
```diff
@@ -1,31 +1,64 @@
...
+
+impl ResolvedDependency {
+    pub(crate) fn package(&self) -> &Package {
+        match self {
+            Self::Remote { package, .. } => package,
+            Self::Local { package, .. } => package,
+        }
+    }
+
+    pub(crate) fn depends_on(&self) -> &Vec<PackageName> {
```
Comment (Tpt) [inline src/resolver.rs:56]: nit: `&[PackageName]` (avoids an extra indirection)

Comment (Tpt) [reply src/resolver.rs:56]:
```suggestion
    pub(crate) fn depends_on(&self) -> &[PackageName] {
```
(avoids an extra indirection)

File: src/registry/cache.rs [hunk 3]
```diff
@@ -35,11 +38,18 @@ impl LocalRegistry {
 
     /// "Downloads" a package from the local filesystem
     pub async fn download(&self, dependency: Dependency) -> miette::Result<Package> {
+        let DependencyManifest::Remote(ref manifest) = dependency.manifest else {
+            return Err(miette!(
+                "unable to serialize version of local dependency ({})",
+                dependency.package
+            ));
+        };
```
Comment (Tpt) [inline src/registry/cache.rs:46]: Nit: not useful code, `dependency_version_string` already does the check

**Review by Tpt — APPROVED** [review]

File: Cargo.toml [hunk 4]
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "buffrs"
-version = "0.8.1"
+version = "0.9.0"
```
Comment (poliorcetics) [inline Cargo.toml:3]: IMO the version bump should be done in another mr to separate concerns

File: src/registry/artifactory.rs [hunk 5]
```diff
@@ -197,6 +200,20 @@ impl Artifactory {
...
+            .dependencies
+            .iter()
+            .find(|d| d.manifest.is_local())
+        {
+            return Err(miette!(
+                "unable to public {} to artifactory due to local dependency {}",
+                package.name(),
+                local.package
+            ));
+        }
```
Comment (poliorcetics) [inline src/registry/artifactory.rs:215]: we could also print the total number of local deps to avoid people forgetting one, getting the error again and restarting the cycle

File: src/resolver.rs (hunk 1 above)
Comment (mara-schulke) [reply src/resolver.rs:42]: Although this is correct im in favor of using types to communicate what is inside of the `String`

File: src/registry/cache.rs (hunk 3 above)
Comment (mara-schulke) [reply src/registry/cache.rs:46]: Yes but that is not technically part of the contract that `dependency_version_string` communicates – ie. although this is the current behavior the check for local dependencies is implicitly happening, rather than explicitly

File: Cargo.toml (hunk 4 above)
Comment (mara-schulke) [reply Cargo.toml:3]: I kind of agree, although this change _is directly tied_ to a manifest change specifically and it should not be rolled out without the version bump

File: src/registry/artifactory.rs (hunk 5 above)
Comment (mara-schulke) [reply src/registry/artifactory.rs:215]: Good hint! Having to rerun buffrs publish over and over just to get a new error every time is a bad ux!
